File: 4_semestre/Paradigmas-de-Linguagem-de-Programacao/aulas/10_outubro/aula_23-10-25/Java-x- C++-x CSharp-GO-Ruby/SistemaBiblioteca/Livro.cs
using System;

namespace SistemaBiblioteca
{
    class Livro
    {
        public string   Titulo      { get; private set; }
        public string   Autor       { get; private set; }
        public bool     Emprestado  { get; private set; }

        public Livro(string titulo, string autor)
        {
            Titulo      = titulo;
            Autor       = autor;
            Emprestado  = false;
        }

        public void Emprestar()
        {
            Emprestado = true;
        }

        public void Devolver()
        {
            Emprestado = false;
        }

        public override string ToString()
        {
            return $"{Titulo} - {Autor}" + (Emprestado ? " (Emprestado)" : " (Disponível)");
        }
    }
}

File: 4_semestre/Paradigmas-de-Linguagem-de-Programacao/paradigmas-de-programação/orientado-a-objetos/paradigma-orientado-objetos-javascript/src/Livro.js
export class Livro {
    constructor(titulo, autor, isbn, anoPublicacao, genero) {
        this._titulo = titulo;
        this._autor = autor;
        this._isbn = isbn;
        this._anoPublicacao = anoPublicacao;
        this._genero = genero;
        this._disponivel = true;
        this._dataEmprestimo = null;
        this._usuarioEmprestimo = null;
    }

    get titulo() {
        return this._titulo;
    }

    get autor() {
        return this._autor;
    }

    get isbn() {
        return this._isbn;
    }

    get disponivel() {
        return this._disponivel;
    }

    emprestar(usuario) {
        if (this._disponivel) {
            this._disponivel = false;
            this._dataEmprestimo = new Date();
            this._usuarioEmprestimo = usuario;
            console.log(`Livro "${this._titulo}" emprestado para ${usuario.nome}.`);
            return true;
        } else {
            console.log(`Livro "${this._titulo}" não está disponível no momento.`);
            return false;
        }
    }

    devolver() {
        if (!this._disponivel) {
            this._disponivel = true;
            const tempoEmprestimo = Math.floor((new Date() - this._dataEmprestimo) / (1000 * 60 * 60 * 24));
            console.log(`Livro "${this._titulo}" devolvido após ${tempoEmprestimo} dias.`);
            this._dataEmprestimo = null;
            this._usuarioEmprestimo = null;
            return true;
        } else {
            console.log(`Livro "${this._titulo}" já está disponível.`);
            return false;
        }
    }

    exibirInfo() {
        return `"${this._titulo}" por ${this._autor} (${this._anoPublicacao}) - ${this._genero} - ISBN: ${this._isbn} - ${this._disponivel ? 'Disponível' : 'Emprestado'}`;
    }
}